internal/policy/verify.go

```diff
@@ -0,0 +1,140 @@
+package policy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/adityasaky/gittuf/internal/gitinterface"
+	"github.com/adityasaky/gittuf/internal/rsl"
+	"github.com/adityasaky/gittuf/internal/tuf"
+	"github.com/go-git/go-git/v5"
+)
+
+var (
+	ErrUnauthorizedSignature = errors.New("unauthorized signature")
+)
+
+func VerifyRef(ctx context.Context, repo *git.Repository, target string) error {
+	// 1. Trace RSL back to the start
+	firstEntryTmp, err := rsl.GetFirstEntry(repo)
+	if err != nil {
+		return err
+	}
+	firstEntry := firstEntryTmp.(*rsl.Entry)
+
+	// 2. Find latest entry for target
+	latestEntryTmp, err := rsl.GetLatestEntryForRef(repo, target)
+	if err != nil {
+		return err
+	}
+	latestEntry := latestEntryTmp.(*rsl.Entry)
+
+	// 3. Do a relative verify from start entry to the latest entry
+	return VerifyRelativeForRef(ctx, repo, firstEntry, latestEntry, target)
+}
+
+func VerifyRelativeForRef(ctx context.Context, repo *git.Repository, firstEntry *rsl.Entry, lastEntry *rsl.Entry, target string) error {
+	entryStack := []*rsl.Entry{lastEntry}
+
+	var currentPolicy *State
+	// 1. Identify policy applicable at firstEntry
+	policyEntry, err := rsl.GetLatestEntryForRefBefore(repo, PolicyRef, firstEntry.ID)
+	if err != nil {
+		return err
+	}
+	state, err := LoadStateForEntry(ctx, repo, policyEntry)
+	if err != nil {
+		return err
+	}
+	currentPolicy = state
+
+	// 2. Enumerate RSL entries between firstEntry and lastEntry, ignoring irrelevant ones
+	iteratorEntry := lastEntry
+	for {
+		if iteratorEntry.GetID() == firstEntry.ID {
+			break
+		}
+
+		parentEntryTmp, err := rsl.GetParentForEntry(repo, iteratorEntry)
+		if err != nil {
+			return err
+		}
+		parentEntry := parentEntryTmp.(*rsl.Entry) // TODO: handle annotations
+
+		if parentEntry.RefName == target || parentEntry.RefName == PolicyRef {
+			entryStack = append(entryStack, parentEntry)
+		}
+
+		iteratorEntry = parentEntry
+	}
+
+	// entryStack has a list of RSL entries in reverse order
+	entryQueue := []*rsl.Entry{}
+	for j := len(entryStack) - 1; j >= 0; j-- {
+		// TODO: instead of reversing it, we can likely just protest this in
+		// reverse order but the policy states are the issue
+		entryQueue = append(entryQueue, entryStack[j])
+	}
+
+	for _, entry := range entryQueue {
+		// FIXME: we're not verifying policy RSL entry signatures because
```

Comment:
This probably deserves a separate issue. Fundamentally, the question is if RSL entries for the policy namespace should be verified for signatures. If we do, we must allow _any_ public key in the delegations tree to make changes to the namespace. In that case, we run into issues such as what key signs an RSL entry for the policy namespace if the changes were to root.json?

On the other hand, we always run `state.Verify` which checks that the policy state is internally consistent, including all reachable metadata signatures starting from root.json. We then have the problem of where we get the root keys from. Currently, it looks into the keys embedded in the repository but these can be easily manipulated if the RSL entry for the policy namespace is not verified. Thoughts? If we use a special case to verify signatures on diffs to root keys / root.json, we again have the question of whether keys used for root.json can also sign git commits for the RSL entry

@JustinCappos @wlynch @znewman01

Reply:
As git supports ssh keys, the last point may not be a problem 🤔